File: UnitTest/Zeze/Trans/TestConcurrentDictionary.cs
using System;
using System.Collections.Generic;
using System.Text;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Collections.Concurrent;

namespace UnitTest.Zeze.Trans
{
    [TestClass]
    public class TestConcurrentDictionary
    {
        [TestMethod]
        public void TestRemoveInForeach()
        {
            ConcurrentDictionary<int, int> cd = new ConcurrentDictionary<int, int>();

            cd.TryAdd(1, 1);
            cd.TryAdd(2, 2);
            cd.TryAdd(3, 3);
            cd.TryAdd(4, 4);
            cd.TryAdd(5, 5);

            int i = 6;
            foreach (var e in cd)
            {
                if (e.Key < 3)
                {
                    cd.TryRemove(e.Key, out var _);
                    Console.WriteLine("remove key=" + e.Key);
                }
                else
                {
                    if (i < 10)
                    {
                        cd.TryAdd(i, i);
                        ++i;
                    }
                    Console.WriteLine("key=" + e.Key);
                }
            }
        }
    }
}

File: ZezeJava/ZezeJavaTest/src/UnitTest/Zeze/Trans/TestConcurrentDictionary.java
package UnitTest.Zeze.Trans;

import junit.framework.TestCase;

public class TestConcurrentDictionary extends TestCase {
	public final void testRemoveInForeach() {
		java.util.concurrent.ConcurrentHashMap<Integer, Integer> cd = new java.util.concurrent.ConcurrentHashMap<>();

		cd.putIfAbsent(1, 1);
		cd.putIfAbsent(2, 2);
		cd.putIfAbsent(3, 3);
		cd.putIfAbsent(4, 4);
		cd.putIfAbsent(5, 5);

		int i = 6;
		for (var e : cd.entrySet()) {
			if (e.getKey() < 3) {
				cd.remove(e.getKey());
				System.out.println("remove key=" + e.getKey());
			} else {
				if (i < 10) {
					cd.putIfAbsent(i, i);
					++i;
				}
				System.out.println("key=" + e.getKey());
			}
		}
	}
}
